From 763a1bfc60f67773cc67f3b42e9fd257f560e9d7 Mon Sep 17 00:00:00 2001
From: Timple Raj M <quic_timple@quicinc.com>
Date: Tue, 30 Jan 2024 20:40:07 +0530
Subject: [PATCH] drivers: clk: enable cpu clk set rate statically

For a low mem profile where CPU_FREQ is disabled,
set cpu freq to a fixed higher rate by forcing it.

This patch is snapshot of the commit
865cc68ea5b1e1a69669f21e86401c5c706c1c40
from win.linuxopenwrt.2.1 branch and
oss/kernel/linux-ipq-5.4 project

Change-Id: Ie6c835c6bd5c3f41163724527ed5003e908a109a
Signed-off-by: Timple Raj M <quic_timple@quicinc.com>
---
 drivers/clk/qcom/apss-ipq-pll.c | 33 ++++++++++++++++++++++++++++++++-
 1 file changed, 32 insertions(+), 1 deletion(-)

diff --git a/drivers/clk/qcom/apss-ipq-pll.c b/drivers/clk/qcom/apss-ipq-pll.c
index 41279e5437a6..9eb189f75f1d 100644
--- a/drivers/clk/qcom/apss-ipq-pll.c
+++ b/drivers/clk/qcom/apss-ipq-pll.c
@@ -8,6 +8,11 @@
 
 #include "clk-alpha-pll.h"
 
+#if !defined(CONFIG_CPU_FREQ) && defined(CONFIG_IPQ_FLASH_16M_PROFILE)
+#define NOMINAL_FREQ	1100000000
+#define TURBO_FREQ	1500000000
+#endif
+
 /*
  * Even though APSS PLL type is of existing one (like Huayra), its offsets
  * are different from the one mentioned in the clk-alpha-pll.c, since the
@@ -168,6 +173,11 @@ static int apss_ipq_pll_probe(struct platform_device *pdev)
 	struct regmap *regmap;
 	void __iomem *base;
 	int ret;
+#if !defined(CONFIG_CPU_FREQ) && defined(CONFIG_IPQ_FLASH_16M_PROFILE)
+	struct clk* cpu_clk;
+	unsigned long rate;
+	struct device_node *np = of_cpu_device_node_get(0);
+#endif
 
 	base = devm_platform_ioremap_resource(pdev, 0);
 	if (IS_ERR(base))
@@ -190,8 +200,29 @@ static int apss_ipq_pll_probe(struct platform_device *pdev)
 	if (ret)
 		return ret;
 
-	return devm_of_clk_add_hw_provider(dev, of_clk_hw_simple_get,
+	ret = devm_of_clk_add_hw_provider(dev, of_clk_hw_simple_get,
 					   &data->pll->clkr.hw);
+	if (ret)
+		return ret;
+
+#if !defined(CONFIG_CPU_FREQ) && defined(CONFIG_IPQ_FLASH_16M_PROFILE)
+	cpu_clk = of_clk_get_by_name(np, "cpu");
+	if (IS_ERR(cpu_clk)) {
+		ret = PTR_ERR(cpu_clk);
+		dev_err(&pdev->dev, "failed to get cpu-clk, %d", ret);
+		return ret;
+	}
+
+	if (cpu_is_ipq5312() || cpu_is_ipq5302())
+		rate = NOMINAL_FREQ;
+	else
+		rate = TURBO_FREQ;
+
+	ret = clk_set_rate(cpu_clk, rate);
+	if (ret)
+		dev_err(&pdev->dev, "failed to set rate for cpu-clk, %d", ret);
+#endif
+	return ret;
 }
 
 static const struct of_device_id apss_ipq_pll_match_table[] = {
-- 
2.34.1

